### PR TITLE
[3pt] One huc processes its own branches

### PR DESCRIPTION
**Pull request by RobHanna-NOAA**
As we move to Amazon Web Service, AWS, we need to change our processing system. Currently, it is `gms_pipeline.sh` using bash "parallel" as an iterator which then first processes all HUCs, but not their branches. One of `gms_pipeline.sh`'s next steps is to do branch processing which is again iterated via "parallel". AKA. Units processed as one step, branches processed as second independent step.

**Note:** While we are taking steps to move to AWS, we will continue to maintain the ability of doing all processing on a single server using a single docker container as we have for a long time. Moving to AWS is simply taking portions of code from FIM and adding it to AWS tools for performance of large scale production runs.

Our new processing system, starting with this PR,  is to allow each HUC to process it's own branches.

A further requirement was to split up the overall processing flow to independent steps, with each step being able to process itself without relying on "export" variables from other files. Note: There are still a few exceptions.  The basic flow now becomes
- `fim_pre_processing.sh`, 
- one or more calls to `fim_process_unit_wb.sh` (calling this file for each single HUC to be processed).
- followed by a call to `fim_post_processing.sh`. 

The new `fim_pipeline.sh` becomes a simple wrapper for the three steps listed above, and when processing multiple HUCs, it uses bash "parallel" as a multi-core HUC processor.

`fim_pipeline.sh`, can be run as a single command taking care of all steps with the same arguments as the previous `gms_pipeline.sh`.  Unlikely `gms_pipeline.sh`, each of the three parts of `fim_pipeline.sh` can be run independently and even as different docker containers if needed as long as they stay in the same order.

### CRITICAL NOTE
The new `fim_pipeline.sh` and by proxy `fim_pre_processing.sh` has two new key input args, one named **-jh** (job HUCs) and one named **-jb** (job branches).  You can assign the number of cores/CPU's are used for processing a HUC versus the number of branches.  For the -jh number arg, it only is used against the `fim_pipeline.sh` file when it is processing more than one HUC or a list of HUCs as it is the iterator for HUCs.   The -jb flag says how many cores/CPU's can be used when processing branches (note.. the average HUC has 26 branches). 

BUT.... you have to be careful not to overload your system.  **You need to multiply the -jh and the -jb values together, but only when using the `fim_pipeline.sh` script.**  Why? _If you have 16 CPU's available on your machine, and you assign -jh as 10 and -jb as 26, you are actually asking for 126 cores (10 x 26) but your machine only has 16 cores._   If you are not using `fim_pipeline.sh` but using the three processing steps independently, then the -jh value has not need to be anything but the number of 1 as each actual HUC can only be processed one at a time. (aka.. no iterator).
</br>
`gms_pipeline.sh` system continues to be available for use as it has been in the past, but will be removed in the near future in favor of exclusive use of `fim_pipeline.sh`.  In order to preserve the current processing system using `gms_pipeline.sh` most files have been created as new files. This results in a bit of duplicate logic in multiple places but this is very temporary as `gms_pipeline.sh` is expected to be removed very soon (possibly the next build).  One exception: For now both the `fim_pipeline.sh` and the `gms_pipeline.sh` both share the individual branch processing file of `src/gms/run_by_branch.sh`

You may notice that a file called `process_unit_wb.sh` exists which creates some "export" global variable values, then calls `run_by_wb.sh` which handles the true processing work.  This "wrapper" type system allows for any errors triggered in `run_by_unit.sh`, to be trapped and managed by `process_unit_wb.sh`. This concept is identical to a previous system using files with names similar to `time_and_tee{}.sh`.  Branch processing follows the same system using `process_branch.sh` and `gms/run_by_branch.sh`.

A number of other minor changes described in the file changes below.

### Additions

- `fim_pipeline.sh` :  The wrapper for the three new major "FIM" processing steps. This script allows processing in one command, same as the current tool of `gms_pipeline.sh`.
- `fim_pre_processing.sh`: This file handles all argument input from the user, validates those inputs and sets up or cleans up folders. It also includes a new system of taking most input parameters and some key enviro variables and writing them out to a files called `runtime_args.env`.  Future processing steps need minimal input arguments as it can read most values it needs from this new `runtime_args.env`. This allows the three major steps to work independently from each other. Someone can now come in, run `fim_pre_processing.sh`, then run `fim_process_unit_wb.sh`, each with one HUC, as many time as they like, each adding just its own HUC folder to the output runtime folder. 
- `fim_post_processing.sh`: Scans all HUC folders inside the runtime folders to handle a number of processing steps which include (to name a few): 
    - aggregating errors
    - aggregating to create a single list (gms_inputs.csv) for all valid HUCs and their branch ids
    - usgs gage aggregation
    - adjustments to SRV's
    - and more    
- `fim_process_unit_wb.sh`: Accepts only input args of runName and HUC number. It then sets up global variable, folders, etc to process just the one HUC. The logic for processing the HUC is in `run_unit_wb.sh` but managed by this `fim_process_unit_wb.sh` file including all error trapping.
- `src`
    - `aggregate_branch_lists.py`:  When each HUC is being processed, it creates it's own .csv file with its branch id's. In post processing we need one master csv list and this file aggregates them. Note: This is a similar file already in the `src/gms` folder but that version operates a bit different and will be deprecated soon.
    - `generate_branch_list.py`: This creates the single .lst for a HUC defining each branch id. With this list, `run_unit_wb.sh` can do a parallelized iteration over each of its branches for processing. Note: This is also similar to the current `src/gms` file of the same name and the gms folder version will also be deprecated soon.
    - `generate_branch_list_csv.py`. As each branch, including branch zero, has processed and if it was successful, it will add to a .csv list in the HUC directory. At the end, it becomes a list of all successful branches. This file will be aggregates with all similar .csv in post processing for future processing.
    - `run_unit_wb.sh`:  The actual HUC processing logic. Note: This is fundamentally the same as the current HUC processing logic that exists currently in `src/gms/run_by_unit.sh`, which will be removed in the very near future. However, at the end of this file, it creates and manages a parallelized iterator for processing each of it's branches.
    - `process_branch.sh`:  Same concept as `process_unit_wb.sh` but this one is for processing a single branch. This file manages the true branch processing file of `src/gms/run_by_branch.sh`.  It is a wrapper file to `src/gms/run_by_branch.sh` and catches all error and copies error files as applicable. This allows the parent processing files to continue despite branch errors. Both the new fim processing system and the older gms processing system currently share the branch processing file of `src/gms/run_by_branch.sh`. When the gms processing file is removed, this file will likely not change, only moved one directory up and be no longer in the `gms` sub-folder.
- `unit_tests`
    - `aggregate_branch_lists_unittests.py' and `aggregate_branch_lists_params.json`  (based on the newer `src` directory edition of `aggregate_branch_lists.py`).
    - `generate_branch_list_unittest.py` and `generate_branch_list_params.json` (based on the newer `src` directory edition of `generate_branch_list.py`).
    -  `generate_branch_list_csv_unittest.py` and `generate_branch_list_csv_params.json` 

### Changes

- `config`
    - `params_template.env`: Removed the `default_max_jobs` value and moved the `startDiv` and `stopDiv` to the `bash_variables.env` file.
    - `deny_gms_unit.lst` : Renamed from `deny_gms_unit_prod.lst`
    - `deny_gms_branches.lst` : Renamed from `deny_gms_branches_prod.lst`

- `gms_pipeline.sh`, `gms_run_branch.sh`, `gms_run_unit.sh`, and `gms_post_processing.sh` :  Changed to hardcode the `default_max_jobs` to the value of 1. (we don't want this to be changed at all). They were also changed for minor adjustments for the `deny` list files names.

- `src`
    - `bash_functions.env`: Fix error with calculating durations.
    - `bash_variables.env`:  Adds the two export lines (stopDiv and startDiv) from `params_template.env`
    - `clip_vectors_to_wbd.py`: Cleaned up some print statements for better output traceability.
    - `check_huc_inputs.py`: Added logic to ensure the file was an .lst file. Other file formats were not be handled correctly.
    - `gms`
        - `delineate_hydros_and_produce_HAND.sh`: Removed all `stopDiv` variable to reduce log and screen output. 
        - `run_by_branch.sh`: Removed an unnecessary test for overriding outputs.

### Removed

- `config`
    - `deny_gms_branches_dev.lst`

### Testing

A wide array of tests where performed against different combination of input params, plus full `fim_pipeline.sh` runs. Tests were also done against each of the three major steps, running some steps a number of times to validate that each step can run 100% independent from any other steps, other than the two .env files on the file system. But also remembering that naturally pre-processing must be done before processing and post processing at the end. For the `process_unit_wb.sh`, tests were performed against one and multiple number of HUCs as well as re-running any one HUC and ensure it affects post-processing as expected.

Test were done against various combination of the new -jh (job HUCs) input arg and the new -jb (job Branches) to ensure how procs/CPU's were being utilized at various points in processing.

Rob Pita and I did an extensive set of testing with many combinations be viewable on a [test matrix.](https://docs.google.com/spreadsheets/d/1WBJmQ5M4sOLCIh-eJliEQm1xjihvPJdaGs5Tq8AXQAU/edit#gid=1858904099)

An alpha domain run will be performed to further validate the timing duration and the quality of the changes. This includes the standard alpha testing (synthesize_test_cases.py and eval_plots.csv).

### CODE REVIEW

File: fim_pipeline.sh [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+set -e
+
+# TODO
+# upgrade Dockerfile to add this as an env value
+projectDir=/foss_fim
+
+echo
+echo "======================= Start of fim_pipeline.sh ========================="
+echo "---- Started: `date -u`" 
+
```
Comment (robgpita-noaa) [inline fim_pipeline.sh:12]: Should we include a `usage` function, or a description of what this script does at a high level?

File: fim_pre_processing.sh [hunk 2]
```diff
@@ -0,0 +1,244 @@
+#!/bin/bash -e
+:
+usage ()
+{
+    echo
+    echo 'Produce FIM hydrofabric datasets for unit and branch scale.'
+    echo 'NOTE: this file runs hucs and branches AND post processing.'
```
Comment (robgpita-noaa) [inline fim_pre_processing.sh:7]: "AND post processing" - isn't that taken care of by `fim_post_processing.sh`?

**Review by robgpita-noaa — COMMENTED** [review]
Nice Work!! Big undertaking to move things over to AWS. Great start here! 

Sorry for the small potatoes comments on `echo` and newlines.  

Couple of initial questions- 

The 3 new `fim_*.sh` scripts are at the top directory - should we place those in `/src`? 

What are your thoughts on updating the file `process_unit_wb.sh` to `fim_process_unit_wb.sh` to keep with the current `fim_` naming convention?

File: src/gms/run_by_branch.sh [hunk 3]
```diff
@@ -23,23 +23,19 @@ outputBranchDataDir=$outputHucDataDir/branches
...
+    rm -rf $outputCurrentBranchDataDir
 fi
 
 ## MAKE OUTPUT BRANCH DIRECTORY
 mkdir -p $outputCurrentBranchDataDir
 
 ## START MESSAGE ##
-echo -e $startDiv"Processing branch_id: $current_branch_id in HUC: $hucNumber ..."$stopDiv
+echo -e $startDiv"Processing HUC: $hucNumber - branch_id: $current_branch_id"
+echo
```
[the remark this answers is not in the document]
Comment (RobHanna-NOAA) [reply src/gms/run_by_branch.sh:35]: ya.. wanted the extra blank line output for readability

File: config/params_template.env [hunk 4]
```diff
@@ -51,9 +51,4 @@ export CALB_DB_KEYS_FILE="/data/config/calb_db_keys.env"
 #### computational parameters ####
 export ncores_gw=1 # mpi number of cores for gagewatershed
 export ncores_fd=1 # mpi number of cores for flow directions
-export default_max_jobs=1 # default number of max concurrent jobs to run
 export memfree=0G # min free memory required to start a new job or keep youngest job alive
-
-#### logging parameters ####
-export startDiv="\n##########################################################################\n"
-export stopDiv="\n##########################################################################"
```
[the remark this answers is not in the document]
Comment (RobHanna-NOAA) [reply config/params_template.env:59]: I find that the new line char doesn't do a thing for it and can likely be removed  everywhere (at least for bash/linux command line outputs) (standard out). But maybe I am wrong. I am planning to get ride of the "stopDiv" export on a near release. It creates extra screen and log output that we don't need.

File: config/params_template.env [hunk 5]
```diff
@@ -51,9 +51,4 @@ export CALB_DB_KEYS_FILE="/data/config/calb_db_keys.env"
 #### computational parameters ####
 export ncores_gw=1 # mpi number of cores for gagewatershed
 export ncores_fd=1 # mpi number of cores for flow directions
-export default_max_jobs=1 # default number of max concurrent jobs to run
```
[the remark this answers is not in the document]
Comment (RobHanna-NOAA) [reply config/params_template.env:54]: I took it out as I never want it to default to more than the value of 1 anywhere and don't want to let that be changeable anywhere, even other bash script or env files.  There are too many accidental by products of it being more than the value of 1 when we are talking multi core and CPU management.

File: fim_pipeline.sh (hunk 1 above)
Comment (RobHanna-NOAA) [reply fim_pipeline.sh:12]: yes and no.  for fim_pipeline, it's usage and "help" is coming via fim_pre_processing.sh" as a proxy.

File: fim_pre_processing.sh (hunk 2 above)
Comment (RobHanna-NOAA) [reply fim_pre_processing.sh:7]: hummm... ya, I need to say that better. I am trying to say that it takes care of fim_post_processing, processing all of the huc(s), then runs fim_post_processing.

File: fim_pipeline.sh (hunk 1 above)
Comment (RobHanna-NOAA) [reply fim_pipeline.sh:12]: I added a quick note above that comment line directing people to the fim_pre_processing.sh top

File: config/params_template.env (hunk 4 above)
[the remark this answers is not in the document]
Comment (robgpita-noaa) [reply config/params_template.env:59]: Was referring to adding a newline at the end of the file.

File: README.md [hunk 6]
```diff
@@ -60,32 +61,34 @@ Input data can be found on the ESIP S3 Bucket (see "Accessing Data through ESIP
...
+Processing of HUC's in FIM4 comes in three pieces. You can run `fim_pipeline.sh` which automatically runs all of three major section, but you can run each of the sections independently if you like. The three sections are:
+- `fim_pre_processing.sh` : This section must be run first as it creates the basic output folder for the run. It also creates a number of key files and folders for the next two sections. 
+- `fim_process_unit_wb.sh` : This script processes one and exactly one HUC8 plus all of it's related branches. While it can only process one, you can run this script multiple times, each with different HUC (or overwriting a HUC). When you run `fim_pipeline.sh`, it automatically iterates when more than one HUC number has been supplied either by command line arguments or via a HUC list. For each HUC provided, `fim_pipeline.sh` will `fim_process_unit_wb.sh`. Using the `fim_proces_unit_wb.sh`  script allows for a run / rerun of a HUC, or running other HUCs at different times / days or even different docker containers.
```
Comment (robgpita-noaa) [inline README.md:79]: `fim_pipeline.sh` will **call** `fim_process_unit_wb.sh`.

File: README.md [hunk 7]
```diff
@@ -60,32 +61,34 @@ Input data can be found on the ESIP S3 Bucket (see "Accessing Data through ESIP
...
-If you choose to do the two step hydrofabric creation, then run `gms_run_unit.sh`, then `gms_run_branch.sh`. See each of those files for details on arguments.
+Running the `fim_pipeline.sh` is a quicker process than running all three steps independently.
 
 ### Testing in Other HUCs
 To test in HUCs other than the provided HUCs, the following processes can be followed to acquire and preprocess additional NHDPlus rasters and vectors. After these steps are run, the "Produce HAND Hydrofabric" step can be run for the new HUCs.
 
 ```
 /foss_fim/src/acquire_and_preprocess_inputs.py -u <huc4s_to_process>
 ```
+    Note: This tool is temporarily un-usable but updates will be coming soon.
```
Comment (robgpita-noaa) [inline README.md:90]: I'd suggest - Note: This tool is deprecated, updates will be coming soon.

File: fim_pipeline.sh [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+- There are a wide number of options and defaulted values, for details run ```gms_pipeline.sh -h```
+- Manditory arguments:
+    - `-u` can be a single huc, a series passed in quotes space delimited, or a line-delimited file
+    i. To run entire domain of available data use one of the ```/data/inputs/included_huc[4,6,8].lst``` files or a huc list file of your choice.
+    - `-n` is a name of your run (only alphanumeric)
+- Outputs can be found under ```/data/outputs/<name_your_run>```
+
+Processing of HUCs in FIM4 (GMS) comes in two pieces: gms_run_unit and gms_run_branch. `gms_pipeline.sh` above takes care of both steps however, you can run each part seperately for faster development if you like.
+
+If you choose to do the two step hydrofabric creation, then run `gms_run_unit.sh`, then `gms_run_branch.sh`. See each of those files for details on arguments.
```
Comment (robgpita-noaa) [inline fim_pipeline.sh:17]: Outdated reference to `gms_run_unit.sh` & `gms_run_branch.sh`

File: fim_pipeline.sh [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+For more details on 
+
+- There are a wide number of options and defaulted values, for details run ```gms_pipeline.sh -h```
+- Manditory arguments:
+    - `-u` can be a single huc, a series passed in quotes space delimited, or a line-delimited file
+    i. To run entire domain of available data use one of the ```/data/inputs/included_huc[4,6,8].lst``` files or a huc list file of your choice.
+    - `-n` is a name of your run (only alphanumeric)
+- Outputs can be found under ```/data/outputs/<name_your_run>```
+
+Processing of HUCs in FIM4 (GMS) comes in two pieces: gms_run_unit and gms_run_branch. `gms_pipeline.sh` above takes care of both steps however, you can run each part seperately for faster development if you like.
```
Comment (robgpita-noaa) [inline fim_pipeline.sh:15]: Outdated references here.

File: README.md (hunk 7 above)
Comment (RobHanna-NOAA) [reply README.md:90]: good idea

Comment (RobHanna-NOAA) [reply README.md:90]: got both (it was mentioned twice)

File: fim_pipeline.sh (hunk 9 above)
Comment (RobHanna-NOAA) [reply fim_pipeline.sh:15]: Fixed.. .thank you (coming in next push)

File: fim_process_unit_wb.sh [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+usage ()
+{
+    echo
+    echo 'Produce FIM hydrofabric datasets for a single unit and branch scale.'
+    echo 'NOTE: fim_pre_processing must have been already run and this tool'
+    echo '      will not include post processing. Only single independent single'
+    echo '      huc and its branches.'    
+    echo 'Usage : There are no arg keys (aka.. no dashes)'
+    echo '        you need the run name first, then the huc.'
+    echo '        ie ) /foss_fim/fim_process_unit_wb.sh rob_test_1 05030104'
```
[the remark this answers is not in the document]
Comment (RobHanna-NOAA) [reply fim_process_unit_wb.sh:28]: Good idea. Implemented. will be on the next commit

File: fim_process_unit_wb.sh [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+# Clean out previous unit logs and branch logs starting with this huc
+rm -f $outputRunDataDir/logs/unit/"$hucNumber"_unit.log
+rm -f $outputRunDataDir/logs/branch/"$hucNumber"_summary_branch.log
+rm -f $outputRunDataDir/logs/branch/"$hucNumber"*.log
+rm -f $outputRunDataDir/unit_errors/"$hucNumber"*.log
+rm -f $outputRunDataDir/branch_errors/"$hucNumber"*.log
+hucLogFileName=$outputRunDataDir/logs/unit/"$hucNumber"_unit.log
+
+# Process the actual huc
+/usr/bin/time -v $srcDir/run_unit_wb.sh | tee $hucLogFileName
```
Comment (robgpita-noaa) [inline fim_process_unit_wb.sh:83]: When submitting a known bad huc to this script directly, the stderr/stdout terminal displays at what point and why the process stopped, however, the stderr/stdout is not making it to the `$hucLogFileName` log. This poses a problem, as when running `fim_pipeline.sh` with a bad huc, the error/exit code are not captured and a user might be under the impression the run was successful. 

`/usr/bin/time -v $srcDir/run_unit_wb.sh | tee $hucLogFileName`
could be replaced with :
`/usr/bin/time -v $srcDir/run_unit_wb.sh 2>&1 | tee $hucLogFileName`
to resolve this. 

See `/foss_fim/outputs/test_robhpr806_fimpreprocessing_ud/logs/unit/04300109_unit.log` for the original run. 
See `/foss_fim/outputs/test_robhpr806_fimpreprocessing_zd/logs/unit/04300109_unit.log` for a run with modification listed above.

File: fim_process_unit_wb.sh [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+        echo
+        echo "***** Unit has no valid branches *****"
+    elif [ $code -eq 61 ]; then
+        echo
+        echo "***** Unit has not a valid unit *****"        
+    else
+        echo
+        echo "***** An error has occured  *****"
+        # copy the error log over to the unit_errors folder to better isolate it
+        cp $hucLogFileName $outputRunDataDir/unit_errors
```
Comment (robgpita-noaa) [inline fim_process_unit_wb.sh:111]: Should we include this line in the two `elif` blocks above as well, so the error is captured and displayed in the `/unit_errors` logs. 

Here is a path to what a log file currently looks like for a bad huc run through `fim_pipeline.sh`: 
`/foss_fim/outputs/test_robhpr806_fimpreprocessing_badhuc/unit_errors/043001009_unit.log`

File: fim_process_unit_wb.sh [hunk 13]
```diff
@@ -0,0 +1,119 @@
+#!/bin/bash -e
+
+# Why is this file here and it appears to be using duplicate export variables?
+# For AWS, we need to make a direct call to this files with two params, hucNumber first, 
+# then the runName same as the -n flag in fim_pipeline and fim_pre_processing.sh
+
+# This file will also catch any and all errors from run_by_huc_wb, even script aborts from that file
```
Comment (robgpita-noaa) [inline fim_process_unit_wb.sh:7]: run_by_huc_wb => `run_unit_wb.sh`?

File: fim_process_unit_wb.sh [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+
+# You really can not call directly to run_unit_wb.sh as that file relys on export values
+# from this file.
+# run_by_huc_wb will futher process branches with its own iterator (parallelization).
```
Comment (robgpita-noaa) [inline fim_process_unit_wb.sh:11]: run_by_huc_wb => `run_unit_wb.sh`?

File: fim_post_processing.sh [hunk 15]
```diff
@@ -0,0 +1,207 @@
+#!/bin/bash -e
+
+# Note: This file does you named arguments as one of them is an optional argument (jobLimit)
```
Comment (robgpita-noaa) [inline fim_post_processing.sh:3]: Providing an overview of what this script accomplishes here would be aweseome! If not, the naming of the script is pretty self-explanatory. I'd suggest either removing or rewording this line. If rewording, something along the lines of: This file requires one named argument (run name), and one optional argument (job limit).

File: fim_process_unit_wb.sh (hunk 11 above)
Comment (RobHanna-NOAA) [reply fim_process_unit_wb.sh:83]: Great idea. Adding it now. Coming on next commit

File: fim_process_unit_wb.sh (hunk 12 above)
Comment (RobHanna-NOAA) [reply fim_process_unit_wb.sh:111]: Yes. I want to add that but decided to skip it for now in favor of letting post processing handle it. I need to clean the unit_errors in certainn conditions and am worried about letting the huc do it and how it handles the non_zero_exit_codes.log against multi core

Comment (robgpita-noaa) [reply fim_process_unit_wb.sh:115]: Sounds great.

**Review by robgpita-noaa — APPROVED** [review]
@RobHanna-NOAA Great work here! My apologies for being a little nit-picky on some things, but thank you for addressing all of the concerns I brought up. 

This PR and its proposed scripts (as well as the ensurance of `gms_pipeline.sh` stability) have been thoroughly tested on the `/dev1` machine at the NWC. 
For folks within the NOAA organization, a test matrix of all of the tests cases that were run can be found in [this google sheet](https://docs.google.com/spreadsheets/d/1WBJmQ5M4sOLCIh-eJliEQm1xjihvPJdaGs5Tq8AXQAU/edit?usp=sharing).

File: fim_pipeline.sh [hunk 16]
```diff
@@ -0,0 +1,76 @@
+#!/bin/bash -e
+
+: '
+fim_pipeline.sh -u <huc8> -n <name_your_run>
+
+For more details on
```
Comment (robgpita-noaa) [inline fim_pipeline.sh:6]: Could this line be omitted?

**Review by robgpita-noaa — DISMISSED** [review]
Looks good! I ran another `fim_pipeline.sh` run with the couple of updates from the commit 9ab4587377f8ac626b4f2cb69e328c2f0742af86 and all checks out!

A couple lines of commented code in `fim_pipeline.sh` could stand to be updated.

File: fim_pipeline.sh (hunk 16 above)
Comment (RobHanna-NOAA) [reply fim_pipeline.sh:6]: I prefer to keep it

**Review by CarsonPruitt-NOAA — APPROVED** [review]

**Review by CarsonPruitt-NOAA — APPROVED** [review]

**Pull request merged**